indicators/Smma/Smma.cs

```diff
@@ -0,0 +1,96 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Skender.Stock.Indicators
+{
+    public static partial class Indicator
+    {
+        // SMOOTHED MOVING AVERAGE
+        /// <include file='./info.xml' path='indicators/type[@name="Main"]/*' />
+        ///
+        public static IEnumerable<SmmaResult> GetSmma<TQuote>(
+            IEnumerable<TQuote> history,
+            int lookbackPeriod)
+            where TQuote : IQuote
+        {
+            // sort history
+            List<TQuote> historyList = history.Sort();
+
+            // check parameter arguments
+            ValidateSmma(history, lookbackPeriod);
+
+            // initialize
+            List<SmmaResult> results = new List<SmmaResult>(historyList.Count);
+
+            // roll through history
+            for (int i = 0; i < historyList.Count; i++)
+            {
+                TQuote h = historyList[i];
+                int index = i + 1;
+
+                SmmaResult result = new SmmaResult
+                {
+                    Date = h.Date
+                };
+
+                // first SMMA calculated as simple SMA
+                if (index == lookbackPeriod)
+                {
+                    decimal sumSma = 0m;
+                    for (int p = index - lookbackPeriod; p < index; p++)
+                    {
+                        TQuote d = historyList[p];
+                        sumSma += d.Close;
+                    }
+
+                    result.Smma = sumSma / lookbackPeriod;
+                }
+                // second SMMA calculated with distinct formula
```

Comment:
I think these two calculations are mathematically equivalent.  I'd drop the third "remaining" calculation in favor of the simpler second one to reduce the code complexity of the algorithm.  The second formula matches what I've seen in other indicators, like ADX and ATR.

Reply:
Started looking at it, refactored a variable to compare the two formulas, this is what I got:

``` csharp
                // second SMMA calculated with distinct formula
                else if (index == lookbackPeriod + 1)
                {
                    decimal prevValue = results[i - 1].Smma.Value;
                    result.Smma = (prevValue * (lookbackPeriod - 1) + h.Close) / lookbackPeriod;
                }
                // remaining SMMA's calculated with base formula
                else if (index >= lookbackPeriod + 2)
                {
                    decimal prevValue = results[i - 1].Smma.Value;
                    result.Smma = ((prevValue * lookbackPeriod) - prevValue + h.Close) / lookbackPeriod;
                }
```

The `index == lookbackPeriod + 1` formula matches up with the Wikipedia formula, but the `index >= lookbackPeriod + 2) formula doesn't seem to match up in a way where I can condense them into one.

Reply:
I'd do it this way, to simplify.

```
            // initialize
            List<SmmaResult> results = new List<SmmaResult>(historyList.Count);
            decimal? prevValue = null;

            // roll through history
            for (int i = 0; i < historyList.Count; i++)
            {
                TQuote h = historyList[i];
                int index = i + 1;

                SmmaResult result = new SmmaResult
                {
                    Date = h.Date
                };

                // calculate SMMA
                if (index > lookbackPeriod)
                {
                    result.Smma = (prevValue * (lookbackPeriod - 1) + h.Close) / lookbackPeriod;
                }

                // first SMMA calculated as simple SMA
                else if (index == lookbackPeriod)
                {
                    decimal sumClose = 0m;
                    for (int p = index - lookbackPeriod; p < index; p++)
                    {
                        TQuote d = historyList[p];
                        sumClose += d.Close;
                    }

                    result.Smma = sumClose / lookbackPeriod;
                }

                prevValue = result.Smma;
                results.Add(result);
            }
```